gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerClusterConfigurationMonitor.java

```diff
@@ -0,0 +1,940 @@
+      }
+    }
+
+    // If service config has not changed, check the role configuration properties
+    if (!hasChanged) {
+      Set<String> previousRoleTypes = previous.getRoleTypes();
+      Set<String> currentRoleTypes = current.getRoleTypes();
+      for (String roleType : previousRoleTypes) {
+        if (!currentRoleTypes.contains(roleType)) {
+          log.roleTypeRemoved(roleType);
+          hasChanged = true;
+          break;
+        } else {
+          previousProps = previous.getRoleProps(roleType);
+          currentProps  = current.getRoleProps(roleType);
+          for (String name : previousProps.stringPropertyNames()) {
+            String prevValue = previousProps.getProperty(name);
+            String currValue = currentProps.getProperty(name);
+            if (currValue == null) { // A missing/removed property
+              if (!(prevValue == null || "null".equals(prevValue))) {
+                log.roleConfigurationPropertyHasChanged(name, prevValue, "null");
+                hasChanged = true;
+                break;
+              }
+            } else  if (!currValue.equals(prevValue)) {
+              log.roleConfigurationPropertyHasChanged(name, prevValue, currValue);
+              hasChanged = true;
+              break;
+            }
+          }
+        }
+      }
+    }
+
+    return hasChanged;
+  }
+
+
+  /**
+   * The thread that polls ClouderaManager for service restart events for clusters associated with discovered
+   * topologies, compares the current service configuration values against the previously-recorded values, and notifies
+   * any listeners when significant configuration differences are discovered.
+   */
+  @SuppressWarnings("PMD.DoNotUseThreads")
+  private static final class PollingConfigAnalyzer implements Runnable {
+
+    private static final int DEFAULT_POLLING_INTERVAL = 60;
+
+    // Polling interval in seconds
+    private int interval;
+
+    private ClouderaManagerClusterConfigurationMonitor monitor;
+
+    private boolean isActive;
+
+    PollingConfigAnalyzer(final ClouderaManagerClusterConfigurationMonitor monitor) {
+      this(monitor, PollingConfigAnalyzer.DEFAULT_POLLING_INTERVAL);
+    }
+
+    PollingConfigAnalyzer(final ClouderaManagerClusterConfigurationMonitor monitor, int interval) {
+      this.monitor   = monitor;
+      this. interval = interval;
+    }
+
+    void setInterval(int interval) {
+      this.interval = interval;
+    }
+
+    void stop() {
+      isActive = false;
+    }
+
+    private void waitFor(long seconds) {
+      try {
+        Thread.sleep(seconds * 1000L);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+
+    @Override
+    public void run() {
+      isActive = true;
+
+      // Initial delay to allow other gateway services (e.g., Topology Service) to be fully ready, so restart events
+      // which occurred while Knox was down are not missed/mishandled.
+      waitFor(20);
```

Comment:
The 20 is what in seconds? This should be waiting on some specific event not just some time.

Reply:
Leveraged the separation of the init() and start() lifecycle events in the gateway services to achieve what I was trying to accomplish.